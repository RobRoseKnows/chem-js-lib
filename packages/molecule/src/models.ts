export enum MoleculeDataFormat {
    jmol = "jmol",
    jnmol = "jnmol",
}
