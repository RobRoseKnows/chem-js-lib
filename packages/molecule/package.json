{
  "name": "@chemistry/molecule-to-svg",
  "version": "2.2.1",
  "description": "Molecule",
  "author": {
    "name": "Volodymyr Vreshch",
    "url": "http://vreshch.com/"
  },
  "license": "MIT",
  "keywords": [
    "cheminformatics",
    "chemistry",
    "molecule",
    "atoms",
    "bonds"
  ],
  "engines": {
    "node": ">=6.0.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/chemistry/chem-js-lib.git"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "private": false,
  "files": [
    "package-lock.json",
    "package.json",
    "dist",
    "README.md"
  ],
  "main": "dist/molecule.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "webpack-cli --config webpack.config.js --mode production"
  },
  "dependencies": {
    "@chemistry/common": "^2.2.0",
    "redux": "4.0.1"
  }
}
