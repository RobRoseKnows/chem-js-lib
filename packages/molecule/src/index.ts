export * from "./molecule";
export * from "./models";
