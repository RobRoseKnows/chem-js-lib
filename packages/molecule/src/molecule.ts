import { JNMol } from "@chemistry/common";
import { createStore } from "redux";
import { MoleculeDataFormat } from "./models";
import {
    exportMolecule,
    exportToSVG,
    IMoleculeState,
    loadMolecule,
    reducer,
} from "./store";

export class Molecule {
    private store = createStore(reducer);

    private get state(): IMoleculeState {
        return this.store.getState();
    }

    public getAtomCount(): number {
        return Object.keys(this.state.atoms).length;
    }

    public getBondCount(): number {
        return Object.keys(this.state.bonds).length;
    }

    public load(data: any, format: MoleculeDataFormat = MoleculeDataFormat.jnmol) {
        this.store.dispatch(loadMolecule(data, format));
    }

    public export(format: MoleculeDataFormat = MoleculeDataFormat.jnmol): any {
        return exportMolecule(this.state, format);
    }

    public toSVG(): JSX.Element {
        return exportToSVG(this.state);
    }
}
