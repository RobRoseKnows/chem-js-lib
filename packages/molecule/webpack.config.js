var webpackConfigFn = require('../../webpack.helper');

module.exports = webpackConfigFn('molecule', __dirname);
