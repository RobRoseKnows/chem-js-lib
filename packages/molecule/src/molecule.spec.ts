import { MoleculeDataFormat } from "./models";
import { Molecule } from "./molecule";

const ACETONE = {
    id: "3",
    title: "acetone",
    atoms: {
         "atom:1": { x: -0.7145, y: -0.4125, z: 0, type: "C" },
         "atom:2": { x: 0, y: 0, z: 0, type: "C" },
         "atom:3": { x: 0.7145, y: -0.4125, z: 0, type: "C" },
         "atom:4": { x: 0, y: 0.825, z: 0, type: "O" },
    },
    bonds: {
        "bond:1": { atom1: "atom:1", atom2: "atom:2", order: 1 },
        "bond:2": { atom1: "atom:2", atom2: "atom:3", order: 1 },
        "bond:3": { atom1: "atom:2", atom2: "atom:4", order: 2 },
    },
};

describe("Molecule", () => {
    let sut: Molecule;

    it("should export class definition", () => {
        expect(Molecule).toBeDefined();
    });

    beforeEach(() => {
        sut = new Molecule();
    });

    it("should be able to create instance", () => {
        expect(sut).toBeDefined();
    });

    describe("JNMol", () => {
        it("should return empty JNMol for new molecule", () => {
            const mol = new Molecule();
            const res = mol.export(MoleculeDataFormat.jnmol);
            expect(res).toEqual({ id: "", title: "", atoms: {}, bonds: {} });
        });
    });

    describe("getAtomCount", () => {
        it("should be able to get atomCount", () => {
            const atomCount = sut.getAtomCount();
            expect(atomCount).toEqual(0);
        });
    });

    describe("getBondCount", () => {
        it("should be able to get getBondCount", () => {
            const bondCount = sut.getBondCount();
            expect(bondCount).toEqual(0);
        });
    });

    describe("load", () => {
        it("should be able to load molecule", () => {
            sut.load(ACETONE, MoleculeDataFormat.jnmol);
            expect(sut.getAtomCount()).toEqual(4);
            expect(sut.getBondCount()).toEqual(3);
        });

        it("should throw for incorrect data format", () => {
            expect(() => {
                sut.load(ACETONE, "unknown-format" as any);
            }).toThrow();
        });

        it("should throw when title is missing in jnmol format", () => {
            expect(() => {
                sut.load({}, MoleculeDataFormat.jnmol);
            }).toThrow();
        });
    });

    describe("noActions", () => {
        it("should ignore unknown actions", () => {
            const state1 = (sut as any).state;
            (sut as any).store.dispatch({ type: "UNKNOWN ACTION "});
            const state2 = (sut as any).state;

            expect(state1).toBe(state2);
        });
    });

    describe("toSVG", () => {
        it("should export ACETONE", () => {
            sut.load(ACETONE, MoleculeDataFormat.jnmol);
            const svg = sut.toSVG();
            expect(svg).toMatchSnapshot();
        });
    });
});
