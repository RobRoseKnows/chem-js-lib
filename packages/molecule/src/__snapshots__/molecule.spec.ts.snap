// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecule toSVG should export ACETONE 1`] = `
<svg>
  <g>
    <text
      alignmentBaseline="middle"
      fill="black"
      textAnchor="middle"
      x={121.42}
      y={33.5}
    >
      C
    </text>
    <text
      alignmentBaseline="middle"
      fill="black"
      textAnchor="middle"
      x={150}
      y={50}
    >
      C
    </text>
    <text
      alignmentBaseline="middle"
      fill="black"
      textAnchor="middle"
      x={178.58}
      y={33.5}
    >
      C
    </text>
    <text
      alignmentBaseline="middle"
      fill="black"
      textAnchor="middle"
      x={150}
      y={83}
    >
      O
    </text>
  </g>
</svg>
`;
